gpflow/expectations.py

```diff
@@ -571,37 +588,33 @@ def _expectation(p, linear_mean, none, kern, feat):
     with params_as_tensors_for(linear_mean):
         N = p.mu.shape[0].value
         D = p.mu.shape[1].value
-        exKxz = expectation(p, mean_functions.Identity(D), (kern, feat))
-        eKxz = expectation(p, (kern, feat))
+        exKxz = expectation(p, mean_functions.Identity(D), (kern, feat), nghp=nghp)
+        eKxz = expectation(p, (kern, feat), nghp=nghp)
         eAxKxz = tf.matmul(tf.tile(linear_mean.A[None, :, :], (N, 1, 1)),
                            exKxz, transpose_a=True)
         ebKxz = linear_mean.b[None, :, None] * eKxz[:, None, :]
         return eAxKxz + ebKxz
 
 
-@dispatch(Gaussian, mean_functions.Constant, type(None), kernels.Kernel, InducingPoints)
-def _expectation(p, constant_mean, none, kern, feat):
+@dispatch(Gaussian, mean_functions.Identity, type(None), kernels.Kernel, InducingPoints)
+def _expectation(p, identity_mean, none, kern, feat, nghp=None):
     """
-    Compute the expectation:
-    expectation[n] = <m(x_n)^T K_{x_n, Z}>_p(x_n)
-        - m(x_i) = c :: Constant function
-        - K_{.,.}    :: Kernel function
-
-    :return: NxQxM
+    This prevents infinite recursion for kernels that don't have specific
+    implementations of _expectation(p, identity_mean, None, kern, feat).
+    Recursion can arise because Identity is a subclass of Linear mean function
+    so _expectation(p, linear_mean, none, kern, feat) would call itself.
+    More specific signatures (e.g. (p, identity_mean, None, RBF, feat)) will
+    be found and used whenever available
     """
-    with params_as_tensors_for(constant_mean):
-        c = constant_mean(p.mu)  # NxQ
-        eKxz = expectation(p, (kern, feat))  # NxM
-
-        return c[..., None] * eKxz[:, None, :]
+    raise NotImplementedError
```

Comment:
So, other cases can't cause infinite loop? Did you check it? Can you prove it?

Reply:
I think it is definitely worth adding @stefanosele's case to the unit tests and show that this is remedied now.

Reply:
Yes I have checked there are no other self-referential calls to `expectation`. Not sure a test is absolutely necessary

Reply:
Ok I added a simple one by including a Matern kernel in exKxz expectation tests

Reply:
I'm happy with that